src/tainting/Dataflow_tainting.ml

```diff
@@ -544,11 +554,28 @@ let handle_taint_propagators env thing taints =
   let propagate_froms, propagate_tos =
     List.partition (fun p -> p.spec.kind =*= `From) propagators
   in
+  (* These are all the labels flowing in to the current thing we're looking at. *)
+  let labels = labels_of_taints taints in
   let lval_env =
     (* `thing` is the source (the "from") of propagation, we add its taints to
      * the environment. *)
     List.fold_left
-      (fun lval_env prop -> Lval_env.propagate_to prop.spec.var taints lval_env)
+      (fun lval_env prop ->
+        (* Only propagate if the current set of taint labels can satisfy the
+           propagator's requires precondition.
+        *)
+        if eval_label_requires ~labels prop.spec.prop.propagator_requires then
+          (* If we have an output label, change the incoming taints to be
+             of the new label.
+             Otherwise, keep them the same.
+          *)
+          let new_taints =
+            match prop.spec.prop.propagator_label with
+            | None -> taints
+            | Some label -> Taints.map (propagate_taint_to_label label) taints
```

Comment:
Now that I look at it again, this actually replaces _all_ labels in `taint` with the propagated `label`. 🤔 Is that what we want to do?
Given:
```yaml
requires: A and B
label: X
```
We could expect `A` and `B` to be replaced with `X`, but if there is also a label `C` in the data... why should it be replaced with `X` ?

Reply:
It could be fine to leave it as it is here for now since it's an experimental feature, but I think this would have to change later.

Reply:
It may be simpler to just have a `replace-labels:` field that needs to be set explicitly?

Reply:
im a little wary of having too many fields, but since it's optional i guess its OK. I've added it now.